src/Microsoft.Azure.ServiceBus/Core/MessageReceiver.cs

```diff
@@ -298,30 +300,46 @@ public Task<IList<Message>> ReceiveAsync(int maxMessageCount)
 
             MessagingEventSource.Log.MessageReceiveStart(this.ClientId, maxMessageCount);
 
+            bool isDiagnosticsEnabled = ServiceBusDiagnosticSource.IsEnabled();
+            Activity activity = isDiagnosticsEnabled ? this.diagnosticSource.ReceiveStart(maxMessageCount) : null;
+            Task receiveTask = null;
+
             IList<Message> unprocessedMessageList = null;
+            IList<Message> processedMessageList = null;
             try
             {
-                await this.RetryPolicy.RunOperation(
+                receiveTask = this.RetryPolicy.RunOperation(
                     async () =>
                     {
-                        unprocessedMessageList = await this.OnReceiveAsync(maxMessageCount, operationTimeout).ConfigureAwait(false);
-                    }, operationTimeout)
-                    .ConfigureAwait(false);
+                        unprocessedMessageList = await this.OnReceiveAsync(maxMessageCount, operationTimeout)
+                            .ConfigureAwait(false);
+                    }, operationTimeout);
+                await receiveTask.ConfigureAwait(false);
+
+                MessagingEventSource.Log.MessageReceiveStop(this.ClientId, unprocessedMessageList?.Count ?? 0);
+
+                if (unprocessedMessageList == null)
+                {
+                    return unprocessedMessageList;
+                }
+
+                processedMessageList = await this.ProcessMessages(unprocessedMessageList).ConfigureAwait(false);
             }
             catch (Exception exception)
             {
                 MessagingEventSource.Log.MessageReceiveException(this.ClientId, exception);
+                if (isDiagnosticsEnabled)
+                {
+                    this.diagnosticSource.ReportException(exception);
+                }
                 throw;
             }
-
-            MessagingEventSource.Log.MessageReceiveStop(this.ClientId, unprocessedMessageList?.Count ?? 0);
-
-            if (unprocessedMessageList == null)
+            finally
             {
-                return unprocessedMessageList;
+                this.diagnosticSource.ReceiveStop(activity, maxMessageCount, receiveTask?.Status, processedMessageList);
```

Comment:
>this.diagnosticSource.ReceiveStop(activity, maxMessageCount, receiveTask?.Status, processedMessageList); [](start = 16, length = 104)

Start-Stop tracking might end up being used as latency counters for simplicity purposes by customers. Lets move the `ReceiveStop` log before `await this.ProcessMessages(..)`